aquadoggo/src/domain.rs

```diff
@@ -0,0 +1,1060 @@
+/// the document id based on operations already existing in the store. Doing this means a document
+/// can be updated without knowing the document id itself.
+///
+/// This method is intended to be used behind a public API and so we assume all passed values
+/// are in themselves valid.
+///
+/// The steps and validation checks this method performs are:
+///
+/// Check if a document view id was passed
+/// - if it wasn't we are creating a new document, safely increment the latest log id for the passed author
+///   and return args immediately
+/// - if it was continue knowing we are updating an existing document
+/// Determine the document id we are concerned with
+/// - verify that all operations in the passed document view id exist in the database
+/// - verify that all operations in the passed document view id are from the same document
+/// - ensure the document is not deleted
+/// Determine next arguments
+/// - get the log id for this author and document id, or if none is found safely increment this authors
+///   latest log id
+/// - get the backlink entry (latest entry for this author and log)
+/// - get the skiplink for this author, log and next seq num
+/// - get the latest seq num for this author and log and safely increment
+/// Return next arguments
+pub async fn next_args<S: StorageProvider>(
+    store: &S,
+    public_key: &Author,
+    document_view_id: Option<&DocumentViewId>,
+) -> Result<NextEntryArguments> {
+    // Init the next args with base default values.
+    let mut next_args = NextEntryArguments {
+        backlink: None,
+        skiplink: None,
+        seq_num: SeqNum::default().into(),
+        log_id: LogId::default().into(),
+    };
+
+    ////////////////////////
+    // HANDLE CREATE CASE //
+    ////////////////////////
+
+    // If no document_view_id is passed then this is a request for publishing a CREATE operation
+    // and we return the args for the next free log by this author.
+    if document_view_id.is_none() {
+        let log_id = next_log_id(store, public_key).await?;
+        next_args.log_id = log_id.into();
+        return Ok(next_args);
+    }
+
+    ///////////////////////////
+    // DETERMINE DOCUMENT ID //
+    ///////////////////////////
+
+    // We can unwrap here as we know document_view_id is some.
+    let document_view_id = document_view_id.unwrap();
+
+    // Get the document_id for this document_view_id. This performs several validation steps (check
+    // method doc string).
+    let document_id = get_validate_document_id_for_view_id(store, document_view_id).await?;
+
+    // Check the document is not deleted.
+    ensure_document_not_deleted(store, &document_id).await?;
+
+    /////////////////////////
+    // DETERMINE NEXT ARGS //
+    /////////////////////////
+
+    // Retrieve the log_id for the found document_id and author.
+    //
+    // @TODO: (lolz, this method is just called `get()`)
+    let log_id = store.get(public_key, &document_id).await?;
```

Comment:
Haha, funny! Why is that? :smile_cat:

Reply:
It's a left over from when we migrated from `aquadoggo` storage methods on dedicated structs, to the more all purpose store :rofl: it's already in the tracking issue for renaming in storage provider: https://github.com/p2panda/aquadoggo/issues/149